lib/Optimizer/GraphOptimizer/GraphOptimizer.cpp

```diff
@@ -1962,56 +1962,147 @@ bool OptimizeReduceMean::run(Function *F, const CompilationContext &cctx) {
+    auto rshpInDims = rshp->getInput().dims();
+    if (rshpInDims.size() != 4) {
+      continue;
+    }
 
-      // Input shape must be 4D.
-      if (RM->getBatch().dims().size() != 4) {
+    auto rshpOutDims = rshp->getResult().dims();
+    if (rshpOutDims.size() != 3) {
+      continue;
+    }
+
+    auto axes = rm->getAxes();
+    if (axes.size() != 1) {
+      continue;
+    }
+
+    // Finally, look for NHWC -> N{HW}C or NCHW -> NC{HW}
+    unsigned_t newAxes[2];
+    if (axes[0] == 1) {
+      // NHWC -> N{HW}C
+      if (rshpInDims[0] != rshpOutDims[0] || rshpInDims[3] != rshpOutDims[2]) {
         continue;
       }
+      newAxes[0] = 1;
+      newAxes[1] = 2;
+    } else if (axes[0] == 2) {
+      // NCHW -> NC{HW}
+      if (rshpInDims[0] != rshpOutDims[0] || rshpInDims[1] != rshpOutDims[1]) {
+        continue;
+      }
+      newAxes[0] = 2;
+      newAxes[1] = 3;
+    } else {
+      continue;
+    }
 
-      // Last two axes must be reduced.
-      auto axes = RM->getAxes();
-      if (axes.size() != 2 || std::count(axes.begin(), axes.end(), 2) != 1 ||
-          std::count(axes.begin(), axes.end(), 3) != 1) {
+    auto rmRes = rm->getResult();
+    auto newRm = F->createBatchedReduceMean(rm->getName(), rmRes.getType(),
+                                            rshp->getInput(), newAxes);
+    rmRes.replaceAllUsesOfWith(newRm->getResult());
+    changed = true;
+  }
+
+  // For each node:
+  for (auto &node : nodes) {
+    if (auto *RM = dyn_cast<BatchedReduceMeanNode>(&node)) {
+
+      if (!canOptReduceMeanWithAvgPool(RM)) {
         continue;
       }
 
       // RM is already shaped to have the required output shape.
       NodeValue in = RM->getBatch();
+      auto axes = RM->getAxes();
+      ShapeVector shapeAxes(axes.begin(), axes.end());
+
+      // Axes must be sorted for correct erase.
+      std::sort(shapeAxes.rbegin(), shapeAxes.rend());
 
-      std::vector<unsigned_t> kernels = {static_cast<unsigned_t>(in.dims()[2]),
-                                         static_cast<unsigned_t>(in.dims()[3])};
       std::vector<unsigned_t> strides = {1, 1};
       std::vector<unsigned_t> pads = {0, 0, 0, 0};
+      std::vector<unsigned_t> kernels = {
+          static_cast<unsigned_t>(in.dims()[shapeAxes[1]]),
+          static_cast<unsigned_t>(in.dims()[shapeAxes[0]])};
 
-      // TODO: Fix bad assumption? See issue 3499, for now workaround it.
       // In Glow, AvgPool expects NHWC.
-      auto *TR1 = F->createTranspose(
-          RM->getName().str() + ".transposeNCHW2NHWC", in, NCHW2NHWC, "NHWC");
-      auto *AP = F->createAvgPool(RM->getName().str() + ".avgPool", TR1,
-                                  kernels, strides, pads);
+      bool needsTranspose = shapeAxes[0] == 3 && shapeAxes[1] == 2; // NCHW?
+
+      // canOptReduceMeanWithAvgPool ensures that shapeAxes satisfies one of
+      // these conditions
+      assert((needsTranspose || shapeAxes[0] == 2 && shapeAxes[1] == 1) &&
```

Comment:
Is there no need to check the opposite ordering for `shapeAxes`, like you do in `canOptReduceMeanWithAvgPool`?

Reply:
`shapeAxes` gets sorted above on line 2092 so it's not needed.